pkg/rsstorage/servers/file/server.go

```diff
@@ -127,28 +139,64 @@ func (s *StorageServer) CalculateUsage() (types.Usage, error) {
 
 // diskUsage will walk the specified path in a filesystem and
 // aggregate the size of the contained files.
-func diskUsage(duPath string, cacheTimeout time.Duration) (size datasize.ByteSize, err error) {
-	timeout := time.Now().Add(cacheTimeout)
-	sizep := &size
+func diskUsage(duPath string, cacheTimeout, walkTimeout time.Duration) (size datasize.ByteSize, err error) {
+	stop := make(chan bool)
+	defer close(stop)
+
+	sizeChan := make(chan datasize.ByteSize)
+	// errChan should have a buffer of two items to prevent deadlock between `<-stop` and `errChan<-err`
+	errChan := make(chan error, 2)
+
+	go func(stop <-chan bool, sizeChan chan<- datasize.ByteSize, errChan chan<- error) {
+		err = filepath.Walk(duPath, func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+
+			select {
+			case <-stop:
+				return nil
+			default:
+			}
+
+			if !info.IsDir() {
+				sizeChan <- datasize.ByteSize(info.Size())
+			}
+
+			return nil
+		})
 
-	err = filepath.Walk(duPath, func(path string, info os.FileInfo, err error) error {
 		if err != nil {
-			return err
+			errChan <- err
 		}
 
-		if time.Now().After(timeout) {
-			return errors.New("timeout in diskUsage")
-		}
+		close(sizeChan)
+		close(errChan)
```

Comment:
This won't change the behavior now, but you might consider using `defer` for these earlier in case we ever add any short-circuit returns.